clients/wallet/src/horizon.rs

```diff
@@ -426,31 +439,69 @@ impl HorizonClient for reqwest::Client {
 		&self,
 		transaction_envelope: TransactionEnvelope,
 		is_public_network: bool,
+		max_retries: u8,
 	) -> Result<TransactionResponse, Error> {
+		let seq_no = transaction_envelope.sequence_number();
 		let transaction_xdr = transaction_envelope.to_base64_xdr();
 		let transaction_xdr = std::str::from_utf8(&transaction_xdr).map_err(Error::Utf8Error)?;
 
-		let base_url = horizon_url(is_public_network);
-		let url = format!("{}/transactions", base_url);
-
 		let params = [("tx", &transaction_xdr)];
-		let response =
-			self.post(url).form(&params).send().await.map_err(Error::HttpFetchingError)?;
-
-		interpret_response::<TransactionResponse>(response).await.map_err(|e| match e {
-			Error::HorizonSubmissionError { title, status, reason, envelope_xdr }
-				if envelope_xdr.is_none() =>
-			{
-				// let's add a transaction envelope, if possible
-				Error::HorizonSubmissionError {
-					title,
-					status,
-					reason,
-					envelope_xdr: Some(transaction_xdr.to_string()),
-				}
-			},
-			other => other,
-		})
+
+		let mut server_error_count = 0;
+		let mut counter = 1;
+
+		loop {
+			let need_fallback = if server_error_count == max_retries {
+				// reset in case the fallback fails but original url could be running again.
+				server_error_count = 0;
+				true
+			} else {
+				false
+			};
+
+			let base_url = horizon_url(is_public_network, need_fallback);
+			let url = format!("{}/transactions", base_url);
+
+			let response =
+				ready(self.post(url).form(&params).send().await.map_err(Error::HttpFetchingError))
+					.and_then(|response| async move {
+						interpret_response::<TransactionResponse>(response).await
+					})
+					.await;
+
+			match response {
+				Err(e) if e.is_recoverable() || e.is_server_error() => {
+					if e.is_server_error() {
+						server_error_count += 1;
+					}
+
+					// let's wait awhile before resubmitting.
+					tracing::warn!(
+						"submission failed for transaction with sequence number {seq_no:?}: {e:?}"
+					);
+					// exponentially sleep before retrying again
+					sleep(Duration::from_secs(2u64.pow(counter))).await;
```

Comment:
Since with the exponential growth, this sleep duration might get huge with enough failures it might make sense to let it saturate at a max value. 
Maybe we can use a min function here. eg. `std::cmp::min(2u64pow(counter), 600)`. We could create another constant for that other value and call it `MAX_BACKOFF_DELAY` or similar.

Reply:
instead of 600, i've set the `MAX_BACKOFF_DELAY` to the power of **_9_**; 2^9 is the nearest to 600.